server/src/main/java/io/crate/expression/predicate/NotPredicate.java

```diff
@@ -112,50 +110,48 @@ void add(Reference symbol) {
+    /**
+     * Three valued logic systems are defined in logic as systems in which there are 3 truth values: true,
+     * false and an indeterminate third value (in our case null is the third value).
+     * <p>
+     * This is a set of cases for which inputs evaluating to null needs can be explicitly excluded
+     * in the boolean queries
+     */
+    private static class SkipThreeValuedLogicVisitor extends SymbolVisitor<SkipThreeValuedLogicContext, Void> {
 
-        /**
-        * Three valued logic systems are defined in logic as systems in which there are 3 truth values: true,
-        * false and an indeterminate third value (in our case null is the third value).
-        * <p>
-        * This is a set of functions for which inputs evaluating to null needs to be explicitly included or
-        * excluded (in the case of 'not ...') in the boolean queries
-        */
-        private final Set<String> STRICT_3VL_FUNCTIONS =
-            Set.of(
-                AnyEqOperator.NAME,
-                AnyNeqOperator.NAME,
-                AndOperator.NAME,
-                AnyRangeOperator.Comparison.GT.opName(),
-                AnyRangeOperator.Comparison.GTE.opName(),
-                AnyRangeOperator.Comparison.LT.opName(),
-                AnyRangeOperator.Comparison.LTE.opName(),
-                LikeOperators.ANY_LIKE,
-                LikeOperators.ANY_NOT_LIKE,
-                CoalesceFunction.NAME,
-                CaseFunction.NAME
-            );
+        private final Set<String> IGNORE_FUNCTIONS = Set.of(ImplicitCastFunction.NAME, ExplicitCastFunction.NAME, MatchPredicate.NAME);
 
         @Override
-        public Void visitReference(Reference symbol, SymbolToNotNullContext context) {
+        public Void visitReference(Reference symbol, SkipThreeValuedLogicContext context) {
             context.add(symbol);
             return null;
         }
 
         @Override
-        public Void visitFunction(Function function, SymbolToNotNullContext context) {
+        public Void visitFunction(Function function, SkipThreeValuedLogicContext context) {
             String functionName = function.name();
             if (Ignore3vlFunction.NAME.equals(functionName)) {
+                context.skip3ValuedLogic = true;
                 return null;
-            }
-            if (!STRICT_3VL_FUNCTIONS.contains(functionName)) {
-                for (Symbol arg : function.arguments()) {
-                    arg.accept(this, context);
+            } else if (EqOperator.NAME.equals(functionName)) {
+                var a = function.arguments().get(0);
+                var b = function.arguments().get(1);
+                // in the  case x = `foo` or foo = `x` 3vl can be skipped
+                if (a instanceof Reference && b instanceof Literal<?> ||
+                    a instanceof Literal<?> && a instanceof Reference) {
+                    context.skip3ValuedLogic = true;
                 }
```

Comment:
Wouldn't this also apply to all other comparison operators?

Reply:
Yes true, that's a good point.

Reply:
Or is there a more general property we could identify that let's us skip 3vl? 

Maybe as part of a Function Feature https://github.com/crate/crate/blob/6f2278fa55d061e576e8f0c4e69ecc340ee4d2b6/server/src/main/java/io/crate/metadata/Scalar.java#L152 ?

E.g. `x = 'foo'` would now get optimized, but `x` = 'fo' || 'o'` wouldn't (unless it's normalized already) ?

I'm a bit worried that we're causing a huge performance regression here for cases that did and could get optimized

Reply:
So you would define it as a function feature that we can skip 3vl ?

Reply:
No, I don't know what I would do. I'm asking if there's a more general aspect that would let us avoid having to list all exceptions. Or at least identify what property a function needs to have to ignore 3vl in NOT

If we do need to provide some information, then we _could_ use the scalar feature properties to communicate that - if it helps.

The initial assumption was that some functions like `coalesce` can recover a `null` into a non-null value. Only a few functions do that.
Another aspect are arrays - where even if the value isn't null, an element might be. 

E.g. one question is why the `NOT (c0||0)<-1` from the latest issue doesn't work. There's no null-recovery for `c0`

Reply:
Let me think if there is a better way to handle this.